Add a job results API for querying LAVA jobs

eds-collabora opened the pull request:
This provides a byte stream of the junit test results for a completed LAVA job, using the standard REST junit endpoint that LAVA provides in v0.2.

The reason to provide a raw bytestream, rather than parsed Junit data, is the difficulty of choosing any particular junit crate for rust at the present: none of them are comprehensive, and it really depends what you want to do with the data which you should pick. It seems safer for now to avoid having any policy here.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

eds-collabora committed:
Add a job results API for querying LAVA jobs

This provides a byte stream of the junit test results for a completed
LAVA job, using the standard REST junit endpoint that LAVA provides
in v0.2.

The reason to provide a raw bytestream, rather than parsed Junit data,
is the difficulty of choosing any particular junit crate for rust at
the present: none of them are comprehensive, and it really depends
what you want to do with the data which you should pick. It seems
safer for now to avoid having any policy here.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

sjoerdsimons commented on lava-api/src/job.rs:
```diff
@@ -538,6 +539,32 @@ pub async fn cancel_job(lava: &Lava, id: i64) -> Result<(), CancellationError> {
+    UnexpectedReply(reqwest::StatusCode),
+}
+
+pub async fn job_results_as_junit(lava: &Lava, id: i64) -> Result<Bytes, ResultsError> {
```
Rather then returning Bytes shouldn't this return a stream instead? As we're not parsing it might make sense to leave it to the user whether to fully read it into memory or not

sjoerdsimons commented:
Could you add a test case for this as well ?

eds-collabora committed:
Permit deserializing Metadata

It's not clear needing to do this represents a design failure or not.
We opted to store the metadata for the TestCases instances in the mock
as strings (even though we have a Metadata type, that we in fact use
to generate that string).

We've now run into the issue that we actually need to know some of the
information in that metadata in order to populate the junit reports that
we would like to return - the failure type and message only exist in the
TestCase Metadata.

There's a more extensive rework possible here where we switch to storing
the metadata as structured, but I believe that this wasn't done because
of the interaction with django - to maintain compatibility with LAVA we
have to treat it as a string, which requires weird carve-outs (i.e. some
kind of serialize-on-demand behaviour when processing queries).

Rather than doing that, the next simplest thing is just to read back
the necessary data from the metadata by deserializing it when we need
it for Junit.  This commit introduces the necessary parts to perform
that deserialization.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

eds-collabora committed:
Add a test for junit support to the lava-api

This is a simple conversion of one of the TestCase unit tests, except
that now we are using the data stored in the mock to double check the
contents of a parsed junit report.

This once again uses junit-parser, but only as a dev-dependency, to
simplify the determination of the contents of the generated report.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

eds-collabora committed:
Add a basic test for mock junit support

This just retrieves four test cases and validates that they have the
expected names and durations. The output is parsed with junit-parser
(a dev-dependency) to limit sensitivity to the surface form of the XML.

This is just a quick smoke test, and not comprehensive; for one thing
all the tests have a success status, and for another there's no check
that the form matches that provided by LAVA for a real report. These
are both matters of time and convenience, in part because serde_xml
has a much more limited set of features than serde_json or serde_yaml.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

eds-collabora committed:
Add support for junit to lava-api-mock

This adds the junit results endpoint for each job to the mocked
API. Rather than introduce new data to back this endpoint, it's driven
by the existing data for TestCases and TestSuites which already exists
in the shared state.

The major wrinkle here is the need to parse the TestCase metadata in
order to report the message and failure type in the generated junit.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

eds-collabora committed:
Use the nested endpoint helper function for matching

There's no need to construct the necessary regular expression match
for a nested endpoint API, when we already have a helper function to
do this in a less error-prone way.

Signed-off-by: Ed Smith <ed.smith@collabora.com>

sjoerdsimons commented on lava-api/src/job.rs:
```diff
@@ -538,6 +539,48 @@ pub async fn cancel_job(lava: &Lava, id: i64) -> Result<(), CancellationError> {
+        .push("junit")
+        .push("");
+
+    Box::pin(
+        async move {
+            let b: Pin<Box<dyn Stream<Item = Result<Bytes, ResultsError>> + Send>> =
+                match lava.client.get(url).send().await {
+                    Ok(res) => match res.status() {
+                        StatusCode::OK => Box::pin(
+                            res.bytes_stream()
+                                .map(|item| item.map_err(ResultsError::from)),
+                        ),
+                        s => Box::pin(stream::once(async move {
+                            Err(ResultsError::UnexpectedReply(s))
+                        })),
+                    },
+                    Err(e) => Box::pin(stream::once(async move { Err(e.into()) })),
+                };
+            b
+        }
+        .flatten_stream(),
+    )
+}
```
That's a heck of  lot of hoops to jump through just becuase `job_results_as_junit` is a sync method :) I've played this out a bit and the above can be collapsed into if the function is async:

```rust
    let res = lava.client.get(url).send().await?;
    match res.status() {
        StatusCode::OK => Ok(res.bytes_stream().map_err(ResultsError::from)),
        s => Err(ResultsError::UnexpectedReply(s)),
    }
```

Which is way more pleasant to read :)

Full patch in: https://github.com/collabora/lava-api/commit/a49922a69a8910d7f309b46a8592face4ffa1253

eds-collabora replied:
I did start with basically what your patch shows, minus the neat `map_err` method from `TryStreamExt` that I was unaware of. I switched to the code above to simplify the callers.

As your patch shows, it's not really about sync vs async. The largest part of these changes is whether you want the errors to be rolled into the stream (simplifying the caller) or to have two different locations for errors to emerge at (immediately, or during stream reading). By rolling the errors in, we permit this to be used in quite naive ways in its callers (i.e. no additional error to map), we're purely returning the thing the caller wants. Alternatively, by separating the errors, sure we can make this code prettier.

Ultimately, it's a judgement call whether you pay the price in the library (once), or in the clients that need it, and while I generally still think this is right trade-off (putting this code here, making the library function interface simpler, at the cost of the details) there's no indisputably correct choice as far as I can see.

By making the change you suggest, we end up with pretty similar code elsewhere (in `lava-gitlab-runner` for one place) to roll the errors into the stream and undo the asyncness. It's not clear whether anywhere else will want to do the same thing, but I think it's arguable that having a pure stream as a return is generally nicer for future clients to work with.

If you're unmoved by this argument ;) then let me know, and I will apply the diffs I made when experimenting with this, and roll it back.

sjoerdsimons replied:
For me there two quite different error topic anyway;  One is "we couldn't even start the stream' and the other is "something went wrong during streaming", which seems more then reasonable to split anyway in an API (Which matches what reqwest does as well!).

So yeah i don't see the point; I don't see it as a simplification of the API in the first place and it requires silly gymnastics to implement which isn't great for maintainability

sjoerdsimons commented on lava-api-mock/src/junit.rs:
```diff
@@ -0,0 +1,98 @@
+fn get_duration(tc: &crate::TestCase<State>) -> Option<Duration> {
+    tc.measurement.as_ref().map(|m| {
+        Duration::seconds_f64(
+            //<crate::testcases::Decimal as Into<rust_decimal::Decimal>>::into(m)
```
Is this meant as a comment or left-over test code?

sjoerdsimons commented on lava-api-mock/src/junit.rs:
```diff
@@ -0,0 +1,98 @@
+    fn respond(&self, request: &Request) -> ResponseTemplate {
+        let rr = Regex::new(r"/api/v0.2/jobs/(?P<parent>[0-9]+)/junit/").unwrap();
+        if let Some(captures) = rr.captures(request.url.as_str()) {
+            println!("Got capture {:?}", captures.get(1).unwrap().as_str());
```
left-over debug?

sjoerdsimons commented:
bors r+
